proto/substrait/algebra.proto

```diff
@@ -594,8 +643,13 @@ message MergeJoinRel {
   RelCommon common = 1;
   Rel left = 2;
   Rel right = 3;
-  repeated Expression.FieldReference left_keys = 4;
-  repeated Expression.FieldReference right_keys = 5;
+  // These fields are deprecated in favor of `keys`.  If they are set then
+  // the two lists (left_keys and right_keys) must have the same length and
+  // the comparion function is considered to be SimpleEqualityType::EQ
+  repeated Expression.FieldReference left_keys = 4 [deprecated = true];
+  repeated Expression.FieldReference right_keys = 5 [deprecated = true];
+  // One or more keys to join on.  The relation is invalid if this is empty.
+  repeated EquiJoinKey keys = 8;
```

Comment:
The fact that we're using EquiJoinKey here makes me want to generalize it further to something like `ComparisonJoinKeys`.

My understanding is that hash joins are restricted to equality-like comparisons for lookup purposes, but for merge joins conditions like `a > b` are perfectly valid. This is representable with what you have now with the `custom_function_reference`, but it would mean putting something that is very much _not_ an equality predicate into a bunch of messages prefixed with `Equality*`.

Reply:
This is correct.  I tossed and turned on this because it's not implemented very consistently at the moment.

In spark's merge join this is an expression, and so it can be anything, even things that don't make sense
In duckdb's merge join this is an expression type, which is overly flexible, similar to spark
In both velox and datafusion the merge join can only do equality conditions

I toyed with the idea of making `EquiJoinKey` not generic (make it part of `HashJoinRel` and revert `MergeJoinRel`.  We could then create a `ComparisonJoinKey` for merge join that, in its list of simple comparison types, includes the basic comparison functions.  But I think this is too much redundancy.  Probably best, if we want to include the simple comparison types (`<`, `<=`, `>`, `>=`) then we go ahead and do that in some future PR and let there be the potential to create an invalid plan (a hash join with `<` for example).  You can already create an invalid plan by choosing a custom comparison operator that either isn't binary and boolean or that doesn't agree with the hash function you use.

This is a very long winded way of saying I agree with your proposal and have switched to it.